app/src/behavior_key_scanned.c

```diff
@@ -0,0 +1,194 @@
+    struct gpio_dt_spec key;
+};
+
+struct behavior_key_scanned_data {
+    struct zmk_behavior_binding binding;
+    struct zmk_debounce_state debounce_state;
+    struct gpio_callback key_callback;
+    const struct device *dev;
+    struct k_work_delayable update_work;
+    uint32_t read_time;
+    bool pin_active;
+    bool active_scan_detected;
+    struct k_sem sem;
+};
+
+static void bks_enable_interrupt(const struct device *dev, bool active_scanning) {
+    const struct behavior_key_scanned_config *config = dev->config;
+
+    gpio_pin_interrupt_configure_dt(&config->key, active_scanning ? GPIO_INT_EDGE_TO_ACTIVE
+                                                                  : GPIO_INT_LEVEL_ACTIVE);
+}
+
+static void bks_disable_interrupt(const struct device *dev) {
+    const struct behavior_key_scanned_config *config = dev->config;
+
+    gpio_pin_interrupt_configure_dt(&config->key, GPIO_INT_DISABLE);
+}
+
+static void bks_read(const struct device *dev) {
+    const struct behavior_key_scanned_config *config = dev->config;
+    struct behavior_key_scanned_data *data = dev->data;
+
+    if (k_sem_take(&data->sem, K_NO_WAIT) < 0) {
+        // k_work_reschedule(&data->update_work, K_NO_WAIT);
+        return;
+    }
+
+    zmk_debounce_update(&data->debounce_state, data->active_scan_detected,
+                        config->debounce_scan_period_ms, &config->debounce_config);
+
+    if (zmk_debounce_get_changed(&data->debounce_state)) {
+        const bool pressed = zmk_debounce_is_pressed(&data->debounce_state);
+
+        struct zmk_behavior_binding_event event = {.position = INT32_MAX,
+                                                   .timestamp = k_uptime_get()};
+
+        if (pressed) {
+            behavior_keymap_binding_pressed(&data->binding, event);
+        } else {
+            behavior_keymap_binding_released(&data->binding, event);
+        }
+    }
+
+    if (zmk_debounce_is_active(&data->debounce_state)) {
+        data->active_scan_detected = false;
+        data->read_time += config->debounce_scan_period_ms;
+
+        k_work_schedule(&data->update_work, K_TIMEOUT_ABS_MS(data->read_time));
+    } else {
+        bks_enable_interrupt(dev, false);
+    }
+
+    k_sem_give(&data->sem);
+}
+
+static void bks_update_work(struct k_work *work) {
+    struct k_work_delayable *dwork = CONTAINER_OF(work, struct k_work_delayable, work);
+    struct behavior_key_scanned_data *data =
+        CONTAINER_OF(dwork, struct behavior_key_scanned_data, update_work);
+    bks_read(data->dev);
+}
+
+static void bks_gpio_irq_callback(const struct device *port, struct gpio_callback *cb,
+                                  const gpio_port_pins_t pin) {
+    struct behavior_key_scanned_data *data =
+        CONTAINER_OF(cb, struct behavior_key_scanned_data, key_callback);
+    const struct behavior_key_scanned_config *config = data->dev->config;
+
+    uint32_t time = k_uptime_get();
+
+    if (k_sem_take(&data->sem, K_MSEC(10)) < 0) {
```

Comment:
I don't think ISRs are allowed to take a semaphore? The documentation implies that ISRs can *give* a semaphore, but only threads can take one.

Reply:
Good catch indeed. Will be pushing a rework that avoids this.